apps/server/src/shared/domain/rules/submission.rule.spec.ts

```diff
@@ -0,0 +1,279 @@
+
+			expect(result).toBe(false);
+		});
+	});
+
+	describe('hasPermission', () => {
+		describe('given user is teacher', () => {
+			describe('given action is "read"', () => {
+				it('should return true if teacher is in course and has required permissions', () => {
+					const course = courseFactory.build({ teachers: [teacher] });
+					const task = taskFactory.build({ course });
+					const submission = submissionFactory.build({ task, student });
+
+					const result = submissionRule.hasPermission(teacher, submission, {
+						action: Actions.read,
+						requiredPermissions: [permissionC],
+					});
+
+					expect(result).toBe(true);
+				});
+
+				it('should return false if teacher does not have required permission', () => {
+					const course = courseFactory.build({ teachers: [teacher] });
+					const task = taskFactory.build({ course });
+					const submission = submissionFactory.build({ task, student });
+
+					const result = submissionRule.hasPermission(teacher, submission, {
+						action: Actions.read,
+						requiredPermissions: [permissionA],
+					});
+
+					expect(result).toBe(false);
+				});
+
+				it('should return false if teacher is not in course', () => {
+					const task = taskFactory.build();
+					const submission = submissionFactory.build({ task, student });
+
+					const result = submissionRule.hasPermission(teacher, submission, {
+						action: Actions.read,
+						requiredPermissions: [],
+					});
+
+					expect(result).toBe(false);
+				});
+			});
+
+			describe('given action is "write"', () => {
+				it('should return true if teacher is in course and has required permissions', () => {
+					const course = courseFactory.build({ teachers: [teacher] });
+					const task = taskFactory.build({ course });
+					const submission = submissionFactory.build({ task, student });
+
+					const result = submissionRule.hasPermission(teacher, submission, {
+						action: Actions.write,
+						requiredPermissions: [permissionC],
+					});
+
+					expect(result).toBe(true);
+				});
+
+				it('should return false if teacher does not have required permission', () => {
+					const course = courseFactory.build({ teachers: [teacher] });
+					const task = taskFactory.build({ course });
+					const submission = submissionFactory.build({ task, student });
+
+					const result = submissionRule.hasPermission(teacher, submission, {
+						action: Actions.write,
+						requiredPermissions: [permissionA],
+					});
+
+					expect(result).toBe(false);
+				});
+
+				it('should return false if teacher is not in course', () => {
+					const task = taskFactory.build();
+					const submission = submissionFactory.build({ task, student });
+
+					const result = submissionRule.hasPermission(teacher, submission, {
+						action: Actions.write,
+						requiredPermissions: [],
+					});
+
+					expect(result).toBe(false);
+				});
+			});
+		});
+
+		describe('given user is student', () => {
```

Comment:
For the rule it should not be different if the user is a student or a teacher. It exist only read / writes. It is possible to add adminstrators as team members for example. 

The task it self has also no fix parent and it should be easy to add new different parents without changing rule code, or tests.

Reply:
I thought about that, too. What we are doing though when composing the test case, is to decide if the user is a teacher or a student. We don't give read or write access explicitly. So I named the tests after the concrete test case and not after the abstract concept.

Reply:
I thougt about it again and I removed the explicit roles from the tests.